### PR TITLE
Hosting: Add visibility of last good backup

### LINKED ISSUE #37258: Hosting: Add visibility of last full site backup and Jetpack Connection Health [2]
Updating files using SFTP or modifying the DB may result in potential data loss. Before making changes, lets help the customer by displaying when the last full site backup occurred.

Here's a mock of how to introduce displaying the last good backup—a new card labeled `Site backup`

**Site backup card:**
Screen Shot 2019-11-14 at 6 47 23 PM

Button links to the activity log, where if I'm reading "I can add enhancement issues for adding the explicit backup request" correctly, they should have a way to initiate a backup once the enhancement issue is worked on. If not, and we're doing it within the SFTP/MySQL page, the button should start the backup.

In page:
Screen Shot 2019-11-14 at 6 47 32 PM

cc @singerb for technical feasibility

cc @apeatling @shaunandrews  @olaolusoga

### PR DESCRIPTION
**Pull request by mmtr**
#### Changes proposed in this Pull Request

Adds a new "Site backup" card to the hosting section that displays the last good backup of the site.

image

To support this, we have added new actions, selectors and reducers to the rewind state. This has required to move the previous rewind state one level deeper, so backups can be stored in the same object:

**Before**
```
console.log( state.rewind );
```
Output:
``` 
{ 
  [siteId]: { 
    state: ..., 
    lastUpdated: ..., 
    ... 
  }
}
```

**After**
```
console.log( state.rewind );
```
Output:
``` 
{ 
  [siteId]: { 
    state: {
      state: ..., 
      lastUpdated: ..., 
      ... 
    },
    backups: [ ... ],
  }
}
```

#### Testing instructions

* Switch to a new Atomic site.
* Go to Manage > Hosting Configuration.
* Verify the Site backup card displays the latest good backup (it should match with what's shown in the activity log).
* Switch to an Atomic site which has not been backed up yet (i.e. a fresh new site).
* Verify the card displays a "There are no recent backups for your site" message.

Fixes #37258

### CODE REVIEW

File: client/my-sites/hosting/site-backup-card/index.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	const lastGoodBackupTime = lastGoodBackup
+		? applySiteOffset( moment.utc( lastGoodBackup.last_updated, 'YYYY-MM-DD hh:mma' ), {
+				timezone,
+				gmtOffset,
+		  } )
+		: null;
+
+	return (
+		<Card className={ classNames( 'site-backup-card', { [ 'is-disabled' ]: disabled } ) }>
+			<CardHeading>{ translate( 'Site backup' ) }</CardHeading>
```
Comment (a8ci18n) [inline client/my-sites/hosting/site-backup-card/index.js]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **28** times:
`translate( 'Site Backups' )` *ES Score: 7*

File: client/state/data-layer/wpcom/sites/rewind/backups/index.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	successRewindBackupsRequest( siteId ),
+	receiveRewindBackups( siteId, backups ),
+];
+
+const displayError = ( { siteId } ) => [
+	failRewindBackupsRequest( siteId ),
+	receiveRewindBackups( siteId, [] ),
+	errorNotice(
+		translate(
+			'Sorry, we had a problem retrieving the site backups. Please refresh the page and try again.'
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/rewind/backups/index.js]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **17** times:
`translate( 'Sorry, we had a problem retrieving your sftp user details. Please refresh the page and try again.' )` *ES Score: 6*

File: client/state/action-types.js [hunk 3]
```diff
@@ -935,6 +935,10 @@ export const REWIND_BACKUP_PROGRESS_REQUEST = 'REWIND_BACKUP_PROGRESS_REQUEST';
 export const REWIND_BACKUP_REQUEST = 'REWIND_BACKUP_REQUEST';
 export const REWIND_BACKUP_UPDATE_ERROR = 'REWIND_BACKUP_UPDATE_ERROR';
 export const REWIND_BACKUP_UPDATE_PROGRESS = 'REWIND_BACKUP_UPDATE_PROGRESS';
+export const REWIND_BACKUPS_RECEIVE = 'REWIND_BACKUPS_RECEIVE';
```
Comment (gwwar) [inline client/state/action-types.js]: Let's avoid the `RECEIVE` naming pattern, it's really not meaningful. Are we adding rewind records here for example?

File: client/state/action-types.js [hunk 4]
```diff
@@ -935,6 +935,10 @@ export const REWIND_BACKUP_PROGRESS_REQUEST = 'REWIND_BACKUP_PROGRESS_REQUEST';
...
+export const REWIND_BACKUPS_REQUEST = 'REWIND_BACKUPS_REQUEST';
+export const REWIND_BACKUPS_REQUEST_FAILURE = 'REWIND_BACKUPS_REQUEST_FAILURE';
+export const REWIND_BACKUPS_REQUEST_SUCCESS = 'REWIND_BACKUPS_REQUEST_SUCCESS';
```
Comment (gwwar) [inline client/state/action-types.js]: Similarly do we need Success and Failure actions?

File: client/state/rewind/backups/reducer.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Returns the updated requests state after an action has been dispatched. The
+ * state maps site ID to whether a request is in progress.
+ *
+ * @param   {object} state  Current state
+ * @param   {object} action Action payload
+ * @returns {object}        Updated state
+ */
+export const requesting = withoutPersistence( ( state = {}, { type } ) => {
```
Comment (gwwar) [inline client/state/rewind/backups/reducer.js]: Do we need to track requesting state? It should be okay to display slightly state data, even if we're in the process of requesting fresh information. The component should be able to infer this if items is equal to the default value of `{}`.

Comment (gwwar) [reply client/state/rewind/backups/reducer.js]: Similarly managing multiple duplicate requests is already handled by the data layer. (Another reason why we used to store this).

**Comment by gwwar** [comment]
@seear did D34701-code or an alternate happen to land?

**Comment by mmtr** [comment]
> @seear did D34701-code or an alternate happen to land?

Yeah, D34701 was abandoned in favor or a new endpoint: https://github.com/Automattic/wp-calypso/issues/37258#issuecomment-555508363

> Update on this. We added a new endpoint instead that lists the last 10 backups along with status:
>
> `GET /wpcom/v2/sites/<site_id>/rewind/backups`
> 
> Source & docs linked from here (first endpoint): Pa0RFL-ct-p2.

File: client/state/action-types.js (hunk 3 above)
Comment (mmtr) [reply client/state/action-types.js]: Renamed to `SET` in 2094e19.

File: client/state/action-types.js (hunk 4 above)
Comment (mmtr) [reply client/state/action-types.js]: Yeah, not needed if we're not tracking the requesting state. Dropped them in f3275c3

File: client/state/rewind/backups/reducer.js (hunk 5 above)
Comment (mmtr) [reply client/state/rewind/backups/reducer.js]: Remove the requesting state in f3275c3

**Comment by mmtr** [comment]
@Automattic/chronos  the `/rewind/backups` endpoint is currently returning a 500 error when the site has just been transferred to atomic. I debugged the `get_backups_list` method and the problems seems to be on `get_backup_attempts` returning an error:

```
(
    [errors] => Array
        (
            [vp_api_error] => Array
                (
                    [0] => Site does not exist.
                )

        )

    [error_data] => Array
        (
        )

)
```

Is this expected? Shouldn't it return an empty list since the site is fresh new and there are no backups?

**Comment by seear** [comment]
> the /rewind/backups endpoint is currently returning a 500 error

500 doesn't seem like a sensible response. @pidgeonman can you take a look at the error handling in the endpoint?

**Comment by mmtr** [comment]
> 500 doesn't seem like a sensible response. @pidgeonman can you take a look at the error handling in the endpoint?

Thanks! In the meantime I added a workaround that fallbacks to the lack of recent backups scenario if the API returns any error.

**Comment by mmtr** [comment]
This is ready for review. Let me know if I should move the data layer changes to a different PR to make easier the review.

**Comment by pidgeonman** [comment]
Here's a PR for sending an empty list on this case https://code.a8c.com/D36031

**Comment by pidgeonman** [comment]
Created diff D36031 for sending an empty list on this case.

**Comment by pidgeonman** [comment]
Changes deployed.

File: client/state/data-layer/wpcom/sites/rewind/backups/index.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+export const setBackups = ( { siteId }, backups ) => setRewindBackups( siteId, backups );
+
+export const resetBackups = ( { siteId } ) => setRewindBackups( siteId, [] );
+
+registerHandlers( 'state/data-layer/wpcom/sites/rewind/backups/index.js', {
+	[ REWIND_BACKUPS_REQUEST ]: [
+		dispatchRequest( {
+			fetch: fetchBackups,
+			onSuccess: setBackups,
+			onError: resetBackups,
```
Comment (gwwar) [inline client/state/data-layer/wpcom/sites/rewind/backups/index.js]: Nice names! ✨

**Comment by seear** [comment]
Screenshot 2019-11-28 at 11 48 02

"You'll lose any content after that date."

This is not true. All Atomic sites back up in realtime using Jetpack Sync, so in theory no data would be lost at all. That date is just the last full snapshot on top of which incremental changes will be applied.

However, I don't think it hurts to be cautious with the messaging.

**Pull request merged**